improver_tests/utilities/test_GradientBetweenVerticalLevels.py

```diff
@@ -0,0 +1,154 @@
+"""Unit tests for the GradientBetweenVerticalLevels plugin."""
+
+
+import iris
+import numpy as np
+import pytest
+
+from improver.synthetic_data.set_up_test_cubes import set_up_variable_cube
+from improver.utilities.gradient_between_vertical_levels import (
+    GradientBetweenVerticalLevels,
+)
+
+
+@pytest.fixture
+def temperature_at_screen_level():
+    """Set up a temperature at screen level cube"""
+    data = np.linspace(0, 3, 4).reshape(2, 2).astype(np.float32)
+    cube = set_up_variable_cube(data)
+    cube.add_aux_coord(iris.coords.AuxCoord(1.5, long_name="height", units="m"))
+    return cube
+
+
+@pytest.fixture
+def temperature_at_850hPa():
+    """Set up a temperature at 850hPa cube"""
+    data = np.linspace(3, 0, 4).reshape(2, 2).astype(np.float32)
+    cube = set_up_variable_cube(data)
+    cube.add_aux_coord(iris.coords.AuxCoord(85000, long_name="pressure", units="Pa"))
+    return cube
+
+
+@pytest.fixture
+def orography():
+    """Set up an orography cube"""
+    data = np.full((2, 2), 8.5, dtype=np.float32)
+    cube = set_up_variable_cube(data, name="surface_altitude", units="m")
+    return cube
+
+
+@pytest.fixture
+def height_of_pressure_levels():
+    """Set up a height of pressure levels cube"""
+    data = np.array([np.full((2, 2), 10), np.full((2, 2), 110)])
+    cube = set_up_variable_cube(
+        data,
+        height_levels=[100000, 85000],
+        pressure=True,
+        name="geopotential_height",
+        units="m",
+    )
+    return cube
+
+
+@pytest.mark.parametrize(
+    "height_or_pressure", ["height_both", "pressure_both", "mixed"]
+)
+def test_height_and_pressure(
+    temperature_at_screen_level,
+    temperature_at_850hPa,
+    orography,
+    height_of_pressure_levels,
+    height_or_pressure,
+):
+    """Test that the plugin produces the expected result with cubes defined either
+    both on height or pressure levels. Also check the plugin produces the expected
+    result with one cube defined on height levels and the other on pressure levels."""
+    cubes = [orography, height_of_pressure_levels]
+    if height_or_pressure == "height_both":
+        temperature_at_850hPa.add_aux_coord(
+            iris.coords.AuxCoord(101.5, long_name="height", units="m")
+        )
+        temperature_at_850hPa.remove_coord("pressure")
+        cubes = []
+    elif height_or_pressure == "pressure_both":
+        temperature_at_screen_level.add_aux_coord(
+            iris.coords.AuxCoord(100000, long_name="pressure", units="Pa")
+        )
+        temperature_at_screen_level.remove_coord("height")
+        cubes = [height_of_pressure_levels]
+
+    cubes.append([temperature_at_850hPa, temperature_at_screen_level])
+
+    expected = [[0.03, 0.01], [-0.01, -0.03]]
+    result = GradientBetweenVerticalLevels()(iris.cube.CubeList(cubes))
+    np.testing.assert_array_almost_equal(result.data, expected)
+    assert result.name() == "gradient_of_air_temperature"
+    assert result.units == "K/m"
```

Comment:
While considering the time metadata, I've discovered that the scalar coordinates are unexpected. In the mixed inputs case, the output has both the height (1.5m) and pressure (85000 Pa) scalar coords while I would expect it to have neither.

The height_both case has no time or forecast_period coord on the output if I change the time of the temperature on pressure levels cube by an hour.

Reply:
I've gone with bounds being added to the case of having just height or just pressure coordinates. Describing the mixed case is a bit more complicated, bounds can't include masked points (the mask is just ignored if you try) and it doesn't make sense to have a coordinate with different units. I've set it up so the cube just contains the height and pressure coordinates of the input cubes to try and descibe what's in the output cube.